master/internal/api_user.go

```diff
@@ -683,3 +696,199 @@ func (a *apiServer) PostUserActivity(
+func (a *apiServer) PostLongLivedToken(
+	ctx context.Context, req *apiv1.PostLongLivedTokenRequest,
+) (*apiv1.PostLongLivedTokenResponse, error) {
+	curUser, _, err := grpcutil.GetUser(ctx)
+	if err != nil {
+		return nil, err
+	}
+	if err = user.AuthZProvider.Get().CanCreateUsersOwnToken(
+		ctx, *curUser); err != nil {
+		return nil, status.Error(codes.PermissionDenied, err.Error())
+	}
+
+	tokenExpiration := user.DefaultTokenLifespan
+	if req.Lifespan != "" {
+		d, err := time.ParseDuration(req.Lifespan)
+		if err != nil || d < 0 {
+			return nil, status.Error(codes.InvalidArgument,
+				"Lifespan must be a Go-formatted duration string with a positive value")
+		}
+		tokenExpiration = d
+	}
+
+	token, err := user.DeleteAndCreateLongLivedToken(
+		ctx, curUser.ID, user.WithTokenExpiry(&tokenExpiration))
+	if err != nil {
+		return nil, err
+	}
+	return &apiv1.PostLongLivedTokenResponse{LongLivedToken: token}, nil
+}
+
+// POST User Long lived token takes user id and optional lifespan and overwrites an
+// access token for the given user.
+func (a *apiServer) PostUserLongLivedToken(
+	ctx context.Context, req *apiv1.PostUserLongLivedTokenRequest,
+) (*apiv1.PostUserLongLivedTokenResponse, error) {
+	curUser, _, err := grpcutil.GetUser(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	targetFullUser, err := getFullModelUser(ctx, model.UserID(req.UserId))
+	if err != nil {
+		return nil, err
+	}
+	targetUser := targetFullUser.ToUser()
+	if err = user.AuthZProvider.Get().CanCreateUsersToken(ctx, *curUser, targetUser); err != nil {
+		return nil, status.Error(codes.PermissionDenied, err.Error())
+	}
+
+	tokenExpiration := user.DefaultTokenLifespan
+	if req.Lifespan != "" {
+		d, err := time.ParseDuration(req.Lifespan)
+		if err != nil || d < 0 {
+			return nil, status.Error(codes.InvalidArgument,
+				"Lifespan must be a Go-formatted duration string with a positive value")
+		}
+		tokenExpiration = d
+	}
+
+	token, err := user.DeleteAndCreateLongLivedToken(
+		ctx, targetFullUser.ID, user.WithTokenExpiry(&tokenExpiration))
+	if err != nil {
+		return nil, err
+	}
+	return &apiv1.PostUserLongLivedTokenResponse{LongLivedToken: token}, nil
+}
+
+// GET Long lived token returns token information for the logged in user.
+func (a *apiServer) GetLongLivedToken(
+	ctx context.Context, req *apiv1.GetLongLivedTokenRequest,
+) (*apiv1.GetLongLivedTokenResponse, error) {
+	curUser, _, err := grpcutil.GetUser(ctx)
+	if err != nil {
+		return nil, err
+	}
+	if err = user.AuthZProvider.Get().CanGetUsersOwnToken(ctx, *curUser); err != nil {
+		return nil, status.Error(codes.PermissionDenied, err.Error())
+	}
+
+	tokenInfo, err := user.GetLongLivedTokenInfo(ctx, curUser.ID)
+	if err != nil {
+		return nil, err
+	}
+
+	return &apiv1.GetLongLivedTokenResponse{LongLivedTokenInfo: tokenInfo.Proto()}, nil
+}
+
+// GET All Long lived token returns all long-lived token information from user_session db.
+func (a *apiServer) GetAllLongLivedToken(
```

Comment:
question (discussion): So currently, we're revoking tokens by deleting them from the table. This might not be the best move for us, since we'd lose out on audit/debugging information regarding historic tokens that were revoked before they expired. 

Instead would we want to add in an additional column for `active` that defaults to true & we filter on when authenticating a session/token? Then a revocation isn't a deletion from the table, but instead an update to that `active` column? The catch with that is that it might also be confusing to have it called `active` in cases where the token has expired & is technically no longer `active`. Alternatively we could have `isRevoked` & have it default to false, so it doesn't get confused with whether or not a token is valid for authentication.

Reply:
Side note: I like that this endpoint doesn't filter down the only active tokens; however, adding pagination would likely a good call since this could potentially get large as time goes on if there's a lot of token use. 

That being said, it should still accept some combination of fields in the request to ensure that we get the full list in the case we're using this from the CLI/SDK.

Reply:
nit (non-blocking): Would we want to have this endpoint be plural, `GetAllLongLivedAccessTokens`?

Reply:
I like the idea of using `isRevoked` column (defaulting to `False`) to mark the DELETE API. 
I will add pagination as well, thanks for pointing it out.

> That being said, it should still accept some combination of fields in the request to ensure that we get the full list in the case we're using this from the CLI/SDK.
What do you mean by combination of fields? For CLI creation, I think we can have the same execution with json output return (similar to swagger)

```
{
  "longLivedTokenInfo": [
    {
      "id": 4979,
      "userId": 2293,
      "expiry": "2024-10-10T02:58:15.092217Z",
      "createdAt": "2024-09-10T02:58:15.092217Z",
      "tokenType": "TOKEN_TYPE_LONG_LIVED_TOKEN",
      "tokenDescription": ""
    },
    {
      "id": 4980,
      "userId": 2294,
      "expiry": "2024-12-09T02:58:27.390935Z",
      "createdAt": "2024-09-10T02:58:27.390935Z",
      "tokenType": "TOKEN_TYPE_LONG_LIVED_TOKEN",
      "tokenDescription": ""
    },
...
  ]
}
```

Reply:
No sweat, sorry I could have been clearer. I was just wanting to make sure even with the addition of pagination we have a way of defining an offset/limit that still allows us to request all the information in one go, since we don't use a paginated version of the data in the CLI/SDK.

Reply:
I have added the isRevoked column, and implemented. But as I have not prototyped CLI yet, so keeping this conversation open.
I will get back to you once I get to the creating CLIs :)

Reply:
Now that I have implemented CLIs, instead of describing the GetAll with pagination through CLI, we agreed of keeping it as a tabular format supporting pagination.

Reply:
Resolving it as it is now implemented